WasmFX-specific documentation in README.md (#200)
wasmfx/wasmfxtime: README.md

This repository contains an up-to-date fork of
[Wasmtime](https://docs.wasmtime.dev/), a standalone WebAssembly engine. This
fork adds support for the [WasmFX](https://wasmfx.dev/) proposal for stack

> switching. 

Reviewer: This repository is a fork of [Wasmtime](https://docs.wasmtime.dev/) extended with support for the [WasmFX](https://wasmfx.dev/) instruction set. WasmFX adds stack switching capabilities to core Wasm, thereby enabling interleaving of computation. The design of the instruction set extension is based on [effect handlers](https://effect-handlers.org), which provide a structured facility for handling non-local control flow.

> 
> The [stack switching repository](https://github.com/WebAssembly/stack-switching) contains a 
> [high-level summary of the proposal](https://github.com/WebAssembly/stack-switching/blob/main/proposals/continuations/Explainer.md),
> [a more formal description](https://github.com/WebAssembly/stack-switching/blob/main/proposals/continuations/Overview.md),
> and [examples](https://github.com/WebAssembly/stack-switching/tree/main/proposals/continuations/examples).

Reviewer: Additional key resources are

```suggestion
* An [explainer document](https://github.com/WebAssembly/stack-switching/blob/main/proposals/continuations/Explainer.md) with an informal introduction to the instruction set extension along with [example programs](https://github.com/WebAssembly/stack-switching/tree/main/proposals/continuations/examples).
* A [formal description](https://github.com/WebAssembly/stack-switching/blob/main/proposals/continuations/Overview.md) of the extension.
* A collection of [benchmark programs](https://github.com/wasmfx/benchfx) and a prototype library for [asynchronous I/O programming](https://github.com/wasmfx/waeio) leveraging the instruction extension.
```

## Building 

The build steps are equivalent to the [standard steps for building Wasmtime from
source](https://docs.wasmtime.dev/contributing-building.html), but using this
repository instead. There is no need to build or install the original version of
Wasmtime to use this fork.

Concretely, the steps are as follows:

1. Make sure that you have a Rust toolchain installed, for example using
   [rustup](https://www.rust-lang.org/tools/install).
2. Check out this repository:

> ``` sh

Reviewer: ```suggestion
```sh
```

git clone https://github.com/wasmfx/wasmfxtime.git
cd wasmfxtime
git submodule update --init
```

> 3. Build:

Reviewer: ```suggestion
3. Build in debug mode:
```

> ``` sh

Reviewer: ```suggestion
```sh
```

cargo build
```

> As a result, a debug build of the `wasmtime` executable will be created at
> `target/debug/wasmtime`.

Reviewer: ```suggestion
To run an artifact, either invoke it via `cargo run` or directly via `target/debug/wasmtime`.
```

> To create a release build instead, run `cargo build --release`, which will
> create `target/release/wasmtime`.

Reviewer: ```suggestion
Analogously, to build in release mode run `cargo build --release`, which places the `wasmtime` artifact in `target/release/`.
```

## Running programs

> A WebAssembly module `my_module.wat` (or `my_module.wasm`) is executed using the
> `wasmtime` executable [in the usual way](https://docs.wasmtime.dev/cli.html). To
> run programs containing WasmFX instructions, enable the necessary features as
> follows:

Reviewer: ```suggestion
The `wasmtime` executable can compile and run WebAssembly modules [in the usual way](https://docs.wasmtime.dev/cli.html). However, if a module contains WasmFX instructions, then it is necessary to enable additional features, e.g.
```

> ``` sh

Reviewer: ```suggestion
```sh
```

> wasmtime -W=exceptions,function-references,typed-continuations my_module.wat

Reviewer: is the `run` subcommand the default command?

Author: Yes

```

> 

Reviewer: ```suggestion

The first two features are prerequisite features, i.e. the `exceptions` feature is required to support WebAssembly tags and the `function-references` feature is required to support typed references. The `typed-continuations` feature enables support for the WasmFX instruction set.

```

> To run an arbitrary function exported as `foo` by the module run 

Reviewer: ```suggestion
To run an arbitrary function exported as `foo` by the module type 
```

``` sh
wasmtime -W=exceptions,function-references,typed-continuations --invoke=foo my_module.wat
```

## Example program

> The following module implements a generator and consumer using stack switching.

Reviewer: I think we should give an entirely self-contained program without any indirection, e.g. "see examples/generator.wat". The reader should be able to copy the example verbatim and run it. I suggest we use an example such as https://dhil.net/research/talks/generator-waw24.wast

Author: The example is self-contained in the sense that you can copy-paste and run it, you just won't get the printed output. The example program you link to relies on the `spectest` module to do the the printing, which is only available when running wasmtime in `wast` mode for running tests. The instructions here are explicitly intended so that you can execute the module in the standard way.

Reviewer: I am not happy with the current example. Maybe we should do a wast example for the output and then go into how to compile and run arbitrary programs. From a usability perspective I think it is important that you can copy paste the first example and observe a concrete output.

Author: I think that we should provide an example that can actually be executed in the same way that you would run an ordinary program. Having to discuss a second way of running programs (namely, in wast mode) seems worse to me than putting the full version of the example behind a link.
Let's see what @slindley thinks once he's back.
